chia/full_node/fee_tracker.py

```diff
@@ -0,0 +1,577 @@
+    ):
+        self.buckets = buckets
+        self.sorted_buckets = sorted_buckets
+        self.confirmed_average = [[] for _ in range(0, max_periods)]
+        self.failed_average = [[] for _ in range(0, max_periods)]
+        self.decay = decay
+        self.scale = scale
+        self.max_confirms = self.scale * len(self.confirmed_average)
+        self.log = log
+        self.fee_store = fee_store
+        self.type = my_type
+        self.max_periods = max_periods
+
+        for i in range(0, max_periods):
+            self.confirmed_average[i] = [0 for _ in range(0, len(buckets))]
+            self.failed_average[i] = [0 for _ in range(0, len(buckets))]
+
+        self.tx_ct_avg = [0 for _ in range(0, len(buckets))]
+        self.m_fee_rate_avg = [0 for _ in range(0, len(buckets))]
+
+        self.unconfirmed_txs = [[] for _ in range(0, self.max_confirms)]
+        for i in range(0, self.max_confirms):
+            self.unconfirmed_txs[i] = [0 for _ in range(0, len(buckets))]
+
+        self.old_unconfirmed_txs = [0 for _ in range(0, len(buckets))]
+
+    def get_bucket_index(self, fee_rate: float) -> int:
+        if fee_rate in self.sorted_buckets:
+            bucket_index = self.sorted_buckets[fee_rate]
+        else:
+            bucket_index = self.sorted_buckets.bisect_left(fee_rate) - 1
+
+        return int(bucket_index)
+
+    def tx_confirmed(self, blocks_to_confirm: int, item: MempoolItem) -> None:
+        if blocks_to_confirm < 1:
+            return
+
+        periods_to_confirm = int((blocks_to_confirm + self.scale - 1) / self.scale)
+
+        fee_rate = item.fee_per_cost * 1000
+        bucket_index = self.get_bucket_index(fee_rate)
+
+        for i in range(periods_to_confirm, len(self.confirmed_average)):
+            self.confirmed_average[i - 1][bucket_index] += 1
+
+        self.tx_ct_avg[bucket_index] += 1
+        self.m_fee_rate_avg[bucket_index] += fee_rate
+
+    def update_moving_averages(self) -> None:
+        for j in range(0, len(self.buckets)):
+            for i in range(0, len(self.confirmed_average)):
+                self.confirmed_average[i][j] *= self.decay
+                self.failed_average[i][j] *= self.decay
+
+            self.tx_ct_avg[j] *= self.decay
+            self.m_fee_rate_avg[j] *= self.decay
+
+    def clear_current(self, block_height: uint32) -> None:
+        for i in range(0, len(self.buckets)):
+            self.old_unconfirmed_txs[i] += self.unconfirmed_txs[block_height % len(self.unconfirmed_txs)][i]
+            self.unconfirmed_txs[block_height % len(self.unconfirmed_txs)][i] = 0
+
+    def new_mempool_tx(self, block_height: uint32, fee_rate: float) -> int:
+        bucket_index: int = self.get_bucket_index(fee_rate)
+        block_index = block_height % len(self.unconfirmed_txs)
+        self.unconfirmed_txs[block_index][bucket_index] += 1
+        return bucket_index
+
+    def remove_tx(self, latest_seen_height: uint32, item: MempoolItem, bucket_index: int) -> None:
```

Comment:
this is not being called anywhere, not sure what the effect on estimation would be

Reply:
Good point - this call in the btc code is the case of "removed from the mempool, but not because it is included in this block". I added it just in case. Thank you.